p2p/protocol/autonatv2/client.go

```diff
@@ -0,0 +1,308 @@
+package autonatv2
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/p2p/protocol/autonatv2/pb"
+	"github.com/libp2p/go-msgio/pbio"
+	ma "github.com/multiformats/go-multiaddr"
+	"golang.org/x/exp/rand"
+)
+
+//go:generate protoc --go_out=. --go_opt=Mpb/autonatv2.proto=./pb pb/autonatv2.proto
+
+// client implements the client for making dial requests for AutoNAT v2. It verifies successful
+// dials and provides an option to send data for dial requests.
+type client struct {
+	host     host.Host
+	dialData []byte
+
+	mu sync.Mutex
+	// dialBackQueues maps nonce to the channel for providing the local multiaddr of the connection
+	// the nonce was received on
+	dialBackQueues map[uint64]chan ma.Multiaddr
+}
+
+func newClient(h host.Host) *client {
+	return &client{host: h, dialData: make([]byte, 4000), dialBackQueues: make(map[uint64]chan ma.Multiaddr)}
+}
+
+// RegisterDialBack registers the client to receive DialBack streams initiated by the server to send the nonce.
+func (ac *client) RegisterDialBack() {
+	ac.host.SetStreamHandler(DialBackProtocol, ac.handleDialBack)
+}
+
+// GetReachability verifies address reachability with a AutoNAT v2 server p.
+func (ac *client) GetReachability(ctx context.Context, p peer.ID, reqs []Request) (Result, error) {
+	ctx, cancel := context.WithTimeout(ctx, streamTimeout)
+	defer cancel()
+
+	s, err := ac.host.NewStream(ctx, p, DialProtocol)
+	if err != nil {
+		return Result{}, fmt.Errorf("open %s stream failed: %w", DialProtocol, err)
+	}
+
+	if err := s.Scope().SetService(ServiceName); err != nil {
+		s.Reset()
+		return Result{}, fmt.Errorf("attach stream %s to service %s failed: %w", DialProtocol, ServiceName, err)
+	}
+
+	if err := s.Scope().ReserveMemory(maxMsgSize, network.ReservationPriorityAlways); err != nil {
+		s.Reset()
+		return Result{}, fmt.Errorf("failed to reserve memory for stream %s: %w", DialProtocol, err)
+	}
+	defer s.Scope().ReleaseMemory(maxMsgSize)
+
+	s.SetDeadline(time.Now().Add(streamTimeout))
+	defer s.Close()
+
+	nonce := rand.Uint64()
+	ch := make(chan ma.Multiaddr, 1)
+	ac.mu.Lock()
+	ac.dialBackQueues[nonce] = ch
+	ac.mu.Unlock()
+	defer func() {
+		ac.mu.Lock()
+		delete(ac.dialBackQueues, nonce)
+		ac.mu.Unlock()
+	}()
+
+	msg := newDialRequest(reqs, nonce)
+	w := pbio.NewDelimitedWriter(s)
+	if err := w.WriteMsg(&msg); err != nil {
+		s.Reset()
+		return Result{}, fmt.Errorf("dial request write failed: %w", err)
+	}
+
+	r := pbio.NewDelimitedReader(s, maxMsgSize)
+	if err := r.ReadMsg(&msg); err != nil {
+		s.Reset()
+		return Result{}, fmt.Errorf("dial msg read failed: %w", err)
+	}
+
+	switch {
+	case msg.GetDialResponse() != nil:
+		break
+	// provide dial data if appropriate
+	case msg.GetDialDataRequest() != nil:
+		idx := int(msg.GetDialDataRequest().AddrIdx)
+		if idx >= len(reqs) { // invalid address index
```

Comment:
Maybe all these checks should go in a `validateDialDataRequest()` method and we can then do s.Reset() once in a single place.